### PR TITLE
Change levels properties size

### PR DESCRIPTION
**Pull request by davidsbailey**
### Background

Today, our curriculum team ran into a problem where they could not copy spritelab code from a spritelab project into the "start code" field on the spritlab level edit page. this caused a panic since it was needed urgently for code break. the team also commented that they have run into this limitation many times before. For additional context, see: https://codedotorg.slack.com/archives/C1B3PNDL7/p1591670561474200

### Description

Increase the maximum size of the levels.properties column to from 16 bits (text) to 24 bits (mediumtext). rails does not let you use the mysql field type, so you have to explicitly set the field length limit to `2^24 - 1` . 

### Rationale

I was worried that the huge increase was somehow irresponsible, so I made a list of potential problems and how we'd deal with them:
1. if we need to efficiently search for any field inside properties json in our production servers, we would typically pull that field out into its own column so we can index it
2. if we need to efficiently search for any field inside properties json for data analysis, we can follow this pattern for creating a separate column for that field in our analysis tables, e.g.: https://github.com/code-dot-org/code-dot-org/blob/9060241bc675bd2175a9ccff6a544eab651d352f/bin/cron/create_rollup_tables#L49
3. if we are worried about this reducing our ability to notice that spritelab source code is getting huge... it really isn't fair to have the way we notice be that curriculum writers are unable to copy spritelab source into start code for a spritelab level. instead, that should be a thing that we investigate separately for the purpose of optimizing the end user experience.

### Verification

I confirmed locally that properties is now `mediumtext` in mysql:
```
mysql> describe levels;
+-----------------------+---------------------+------+-----+---------+----------------+
| Field                 | Type                | Null | Key | Default | Extra          |
+-----------------------+---------------------+------+-----+---------+----------------+
| id                    | int(11)             | NO   | PRI | NULL    | auto_increment |
| game_id               | int(11)             | YES  | MUL | NULL    |                |
| name                  | varchar(255)        | NO   | MUL | NULL    |                |
| created_at            | datetime            | YES  |     | NULL    |                |
| updated_at            | datetime            | YES  |     | NULL    |                |
| level_num             | varchar(255)        | YES  |     | NULL    |                |
| ideal_level_source_id | bigint(20) unsigned | YES  |     | NULL    |                |
| user_id               | int(11)             | YES  |     | NULL    |                |
| properties            | mediumtext          | YES  |     | NULL    |                |
| type                  | varchar(255)        | YES  |     | NULL    |                |
| md5                   | varchar(255)        | YES  |     | NULL    |                |
| published             | tinyint(1)          | NO   |     | 0       |                |
| notes                 | text                | YES  |     | NULL    |                |
| audit_log             | text                | YES  |     | NULL    |                |
+-----------------------+---------------------+------+-----+---------+----------------+
```
I also verified locally that running the migration `up` and then `down` results in no net changes to the db, schema or model files.

### CODE REVIEW

File: dashboard/app/models/levels/go_beyond.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #  level_num             :string(255)
 #  ideal_level_source_id :integer          unsigned
 #  user_id               :integer
-#  properties            :text(65535)
+#  properties            :text(16777215)
```
Comment (dmcavoy) [inline dashboard/app/models/levels/go_beyond.rb:13]: Is this a level type we still use?

File: dashboard/app/models/levels/script_completion.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #  level_num             :string(255)
 #  ideal_level_source_id :integer          unsigned
 #  user_id               :integer
-#  properties            :text(65535)
+#  properties            :text(16777215)
```
Comment (dmcavoy) [inline dashboard/app/models/levels/script_completion.rb:13]: Is this level type still used?

**Review by dmcavoy — APPROVED** [review]
Looks good. Left two comments about level types that I'm wondering if we still use but definitely could be handled in another PR.

**Review by uponthesun — COMMENTED** [review]
Can you test this out on a production clone first? The attempt to add the index a little while back has made me want to err more towards caution.

**Comment by davidsbailey** [comment]
@uponthesun sure. Any docs you can point me to on that?

File: dashboard/app/models/levels/go_beyond.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/models/levels/go_beyond.rb:13]: nope! will remove.

File: dashboard/app/models/levels/script_completion.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/app/models/levels/script_completion.rb:13]: there is no remaining level file, but the level still exists in levelbuilder, prod, and possibly other environments. seems worth cleaning up.

**Comment by uponthesun** [comment]
Docs for creating the clone are here: https://github.com/code-dot-org/code-dot-org/blob/staging/lib/rake/rds.rake

Configuring an adhoc to use a clone is tricky and I don't know if we have docs on this. The hacky way I've done it is to make changes locally on the adhoc to https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/config/database.yml. @sureshc has done the most work on this recently I think.

**Comment by davidsbailey** [comment]
OK, the migration succeeds when running on a clone of production. I started to argue against this, because the source of truth for levels in every environment is the .level files, which would suggest that the migration would work the same in every environment. but, given the way things are made, there are probably some levels which exist only in production, since they could have been deleted from the levelbuilder machine, and then been wiped out by various DB re-installations in staging and test.

**Comment by uponthesun** [comment]
@davidsbailey Thanks for following through with this. How costly was the process of testing this in the end? Do you think it's reasonable to test every migration this way?

**Comment by davidsbailey** [comment]
Good question. :-)

It seemed like it took about 90 minutes to figure out what to do and then do all the steps. It might be faster next time, but there are a lot of steps, not necessarily that easy to remember, a few rough spots in the "instructions" (no special characters in lastpass pwd, no existing db name in globals.yml) and a lot of jumping around between different places to have to do things. 

based on the effort required today, I would only recommend this process if at least two of the following are true:

1. high likelihood of production outage. e.g. migration affects many rows, could take a very long time to run, could lock up db rows, causing 500x errors in our servers.

2. high likelihood of a problem due to blocking our deploy pipeline, perhaps due to a combination of:

    a. high chance of succeeding in one environment and then failing later in a later one:  development --> staging --> test --> production (--> levelbuilder). the later it fails, the more places there are to clean up.

    b. the migration itself is hard to undo. Any migration where bringing the migration `down` effectively undoes bringing the migration `up` would not meet this criteria. Perhaps we learned that adding foreign keys is hard to undo? Perhaps the steps for fixing this in the "easy" case need to be documented?

    c. the migration takes place during a critical time window when we need to keep the pipeline clear (CSP AP deadline, approaching HoC, approaching a carefully-timed deploy, etc)

If we though this step was worth requiring in certain situations, I'd suggest trying to automate it. most of the steps seem like they could be done via a rake task. Perhaps it could then be required/recommended if certain criteria above were met?

**Pull request merged**